include/circt/Dialect/FIRRTL/FIRRTLOps.h

```diff
@@ -174,6 +175,99 @@ struct FirMemory {
   }
   StringAttr getFirMemoryName() const;
 };
+
+// Record of the inner sym name, the module name and the corresponding
+// operation. Also the port index, if the symbol is on a module port.
+struct InnerRefRecord {
+  mlir::StringAttr mod, innerSym;
+  mlir::Operation *op = nullptr;
+  unsigned portIdx = 0;
+  InnerRefRecord(StringAttr mod, StringAttr innerSym, Operation *op)
+      : mod(mod), innerSym(innerSym), op(op) {}
+  InnerRefRecord(StringAttr mod, StringAttr innerSym, Operation *op,
+                 unsigned portIdx)
+      : mod(mod), innerSym(innerSym), op(op), portIdx(portIdx) {}
+  InnerRefRecord(hw::InnerRefAttr ref)
+      : mod(ref.getModule()), innerSym(ref.getName()) {}
+  bool operator<(const InnerRefRecord &rhs) const {
+    return (innerSym.getValue() < rhs.innerSym.getValue() ||
+            (innerSym == rhs.innerSym && mod.getValue() < rhs.mod));
+  }
+  bool operator==(const InnerRefRecord &rhs) const {
+    return (innerSym == rhs.innerSym && mod == rhs.mod);
+  }
+};
+// A list of inner sym and the corresponding operations. Can be used for
+// efficient insertion and then searching. It can be used when the pattern is
+// first all the records are inserted, and then searched for. After insertion,
+// the list must be sorted, and then binary search is used.
+struct InnerRefList {
```

Comment:
I think you might be able to do something more like a hashmap instead of a sorted list + binary search. You might even be able to do something like the following, but I haven't thought it through yet.  It uses the new `AnnoTarget` thing I [merged last week](https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h#L381)

```c++
struct InnerRefRecord {
  StringAttr mod;
  StringAttr innerSym;
}
inline ::llvm::hash_code hash_value(const InnerRefRecord &arg) {
  using AttrPairT = std::pair<Attribute, Attribute>;
  return DenseMapInfo<AttrPairT>::getHashValue(AttrPairT(arg.mod, arg.innerSym));
}

using InnerRefMap = DenseMap<InnerRefRecord, AnnoTarget>;
```

Reply:
I will land this PR for now, then measure any performance difference between Hashmaps vs this.